Link: add optional human recognizable link names

SamuelDudley opened the pull request:
#### EDITED 07/12/2017
When using multiple master links it can become confusing (at least for me!) as to what hardware (RFD, 3G,  Microhard, VPN, Iridium, WIFI, etc..) is attached to each link number (Link 1, 2, 3, etc...). This PR adds an optional label to a link which replaces the link number in the console.

Example of adding a master link called 'RFD900x': `--master=/dev/ttyUSB0:'{"label":"RFD900x"}'`
Example of removing a link called 'RFD900x': `link remove RFD900x`
Example of adding a link called 'FTDI': `link add /dev/ttyUSB0:{"label":"FTDI"}`

_Notes_:

- You can still use the link number to remove links if you don't want to type the label.
- Typing `link list` will show the link number and associated label (if any)

![console](https://cloud.githubusercontent.com/assets/8290680/24227584/11b17c5a-0fbe-11e7-9f7b-72d7e047dbbe.png)

If no json dict with a label key is passed when creating the link it will use the existing link numbering system, making the PR backwards compatible. I'm open to suggestions to improve the implementation and syntax if required.
Cheers,
Sam

tridge commented:
nice idea!
Could you put the repetitive code (the if statements) in a common place, perhaps in mp_module.py?

SamuelDudley commented:
@tridge No problem! I have removed the duplicated code and implemented a static method in mp_module.py. As the static method functionality is only used to access the function from mavproxy.py, perhaps the if statement should remain here? or maybe there is a cleaner way to obtain the same functionality without requiring a static method?

Let me know if your thoughts and if you would like any more changes.
Cheers,
Sam

SamuelDudley commented on MAVProxy/modules/mavproxy_link.py:
```diff
@@ -110,23 +110,29 @@ def show_link(self):
         for i in range(len(self.mpstate.mav_master)):
             conn = self.mpstate.mav_master[i]
-            print("%u: %s" % (i, conn.address))
+            if conn.handle is not None:
```
I have left this as an if statement because I want to be explicit about what data is being (or not being) displayed

SamuelDudley commented:
I feel that this is good to merge. Please let me know if you would like any other changes.
Cheers!

peterbarker commented on MAVProxy/modules/mavproxy_link.py:
```diff
@@ -110,23 +110,29 @@ def show_link(self):
+    def link_add(self, descriptor):
         '''add new link'''
         try:
+            link_components = descriptor.split('(')
```
I wonder if we should do something more comprehensive in terms of adding attributes to links here?  Perhaps name-value pairs?

At the moment you're putting stuff in prens which is being dealt with by MAVProxy.  Another option would be to extend the link strings which pymavlink understands and have it remember the link names.  So your link might be:

`link add /dev/ttyUSB0:label=rfd900z`

peterbarker commented on MAVProxy/modules/mavproxy_link.py:
```diff
@@ -137,11 +143,16 @@ def link_add(self, device):
+        try:
+            handle = link_components[1]
+            handle = handle.rstrip(')')
+            conn.handle = handle
```
Perhaps "label" would be better than "handle"; less confusing with a "file handle" that way...

SamuelDudley commented:
Hi,
Thanks for the feedback :+1: I agree with the re-naming from "handle" to "label".
I'm on the fence about where the string parsing should be located but I'll move to the name value pair arrangement. If we like it as an idea we can pull the parsing from here and insert it into mavutil.py

SamuelDudley commented:
@peterbarker Please have a look at the changes when you have some time and let me know your thoughts. Cheers.

SamuelDudley commented on MAVProxy/modules/mavproxy_link.py:
```diff
@@ -147,6 +161,11 @@ def link_add(self, device):
         conn.last_message = 0
         conn.highest_msec = 0
+        for attr in optional_attributes:
+            if not hasattr(conn, attr):
```
There might be a reason you would want to do this... possibly to do something funky like set target_system here on a per link basis. However to do this we would need to convert the 'value' string into the correct type (int in the case of target_system). This could be done by passing in a single dict of key, value pairs and parsing the dict... /dev/ttyUSB0:{"target_system":123,"label":"link_for_system_123"}

peterbarker commented:
On Sun, 9 Jul 2017, Samuel Dudley wrote:

> Thanks for the feedback 👍 I agree with the re-naming from "handle" to
> "label".
> I'm on the fence about where the string parsing should be located but I'll
> move to the name value pair arrangement. If we like it as an idea we can
> pull the parsing from here and insert it into mavutil.py

I like it!

Perhaps @stephendade could comment?

stephendade commented:
Works nicely on Windows! I vote for merging.

SamuelDudley commented:
Hi, just bumping this PR. I would still very much like to see this merged in the future (when time permits). I am still willing to make any further changes as requested in order to have it merged.
Thanks again,
Sam

SamuelDudley commented:
modified to support json format and rebased off master

SamuelDudley commented:
merged with #480 
Thanks to everyone who helped get this in!
